bifacial_radiance/spectral_utils.py

```diff
@@ -0,0 +1,169 @@
+    
+    interpolation_methods = {
+        'linear':   _linear_interpolation,
+        'nearest':  _nearest_interpolation,
+        'lower':    _lower_interpolation,
+        'upper':    _upper_interpolation
+    }
+    
+    def __init__(self, values, index, index_units='nm', interpolation=None):
+        # Verify lengths match
+        if len(values) != len(index):
+            print("Warning: Length of values and index must match.")
+            return
+        
+        # Convert inputs to list
+        values = [ val for val in values ]
+        index = [ spectral_property.to_nm(idx, index_units) for idx in index ]
+        
+        # Create DataFrame
+        self.df = pd.DataFrame()
+        self.df['value'] = values
+        self.df['wavelength'] = index
+        self.df = self.df.set_index('wavelength')
+        
+        self.interpolation = None
+        if interpolation in spectral_property.interpolation_methods:
+            self.interpolation = \
+                    spectral_property.interpolation_methods[interpolation]
+            self.interpolation_type = interpolation
+        elif interpolation:
+            print("Warning: Specified interpolation type unknown.")
+            
+    def _get_single(self, wavelength, units):
+        # Convert wavelength to nm
+        wavelength = spectral_property.to_nm(wavelength, units)
+        
+        if wavelength in self.df.index:
+            # If the value for that wavelength is known, return it
+            return self.df['value'][wavelength]
+        elif self.interpolation:
+            # Check wavelength is within range
+            if wavelength < self.df.index.min() or \
+               wavelength > self.df.index.max():
+                print("Warning: Requested wavelength outside spectrum.")
+                return None
+            
+            # Return interpolated value
+            return self.interpolation(self, wavelength)
+        
+        return None
+    
+    def __getitem__(self, wavelength, units='nm'):
+        if isinstance(wavelength, Iterable):
+            return np.array([ self._get_single(wl, units) for wl in wavelength ])
+        return self._get_single(wavelength, units)
+    
+    def to_file(self, filepath, append=False):
+        mode = 'w'
+        if append:
+            mode = 'a'
+            
+        with open(filepath, mode) as outfile:
+            outfile.write(f"interpolation:{self.interpolation_type}\n")
+            self.df.to_csv(outfile)
+    
+    def range(self):
+        # Find upper and lower index
+        upper_bound = self.df.index.max()
+        lower_bound = self.df.index.min()
+        
+        return (lower_bound, upper_bound)
+    
+    def scale_values(self, scaling_factor):
+        self.df['value'] *= scaling_factor
+    
+def spectral_albedo_smarts(zen, azm, material, min_wavelength=300,
+                           max_wavelength=4000):
+    try:
+        from pySMARTS.smarts import SMARTSSpectraZenAzm
```

Comment:
@shirubana I was snooping around to see what's going on in bifacial_radiance-land and am interested in this `pySMARTS` package.  I assume it's a python wrapper around SMARTS, but could only find [this](https://github.com/BYU-PRISM/hale-trajectory) after some quick googling.  What's the story on `pySMARTS`?

Reply:
https://github.com/NREL/py-SMARTS

hmmm... story is, we had a matlab wrapper back at my Univeristy; I got permission from the original grad student, advisor, and made some updates then and talked to Cliff about publishing it in pvlib, but never got around that. Then ported it to python when at NREL, made a pull request on PVLib that I've so far have not done much about and then contacted NREL legal to make sure we were doing things right, which involved having our own Github, copyright notes on all of this situation, and from that github make a PR to pvlib.  I'm just not very good at pvlib whole contributing process and it's been not on my priority (or theirs). But it's there, and it does stuff. I've been meaning to take a fun-work day and do a GUI for my own self, I keep having to go read the manual to understand the inputs and it annoys me. .

Reply:
also that zen-azimuth function got updated today and I got a better function to work with TMY and SRRL data with a lot more detail. I have to think how to do the inputs mapping more friendly depending on the data inputs available. But working as need hehe.

Reply:
Thanks!  I think I was searching google for `"pySMARTS"` without the hyphen, so that's why I missed it.